lib/recursive-methods.rb

```diff
@@ -1,49 +1,161 @@
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    str =''
+    index = s.length-1
+    str= reverse_helper(s,str,index)
+    return str
+
 end
 
-# Time complexity: ?
-# Space complexity: ?
+    def reverse_helper(s,str,index)
+        if str.length == s.length
+            return str
+        else 
+            str+= s[index]
+            reverse_helper(s,str,index-1)
+        end
+        
+    end
+
+# Time complexity: O(n) n is s.length
+# Space complexity: O(1)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    b = 0
+    e = s.length - 1
+    reverse_inplace_helper(b,e,s)
+    return s
+end
+    
+    def reverse_inplace_helper(b,e,s)
+    if s == '' || b > s.length/2
+        return s
+    else 
+        temp = s[b]
+        s[b] = s[e]
+        s[e] = temp
+        reverse_inplace_helper(b+1,e-1,s)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0 
+      raise ArgumentError, "bunnies cannot be a negative number!"
+    elsif
+      n == 0
+      return n
+    else
+      return 2+ bunny(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) n is s.length
+# Space complexity: O(1)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    x = 0
+    if s == ''
+        return true
+    elsif
+        s.length%2 != 0
+        return false
+    else
+        return nested_helper(s,x)
+    end
+       
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def nested_helper(s,x)
+    if x >= s.length/2
+        return true
+    elsif  s[x]== "(" && s[s.length-1-x] == ")"
+        nested_helper(s,x+1) 
+    else 
+        return false
+    end  
+end 
+
+# Time complexity: O(n) n is array.length
+# Space complexity: O(1)
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+    index = 0
+    if array == []
+        return false
+    else search_helper(array,value,index)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+   def search_helper(array,value,index)
+    if array[index] == value
+        return true
+    elsif index != array.length-1
+        search_helper(array,value,index+1)
+    else
+        return false
+    end
+   end
+
+# Time complexity: O(n) s.length
+# Space complexity: O(1)
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+    index = 0
+    if s == ''
+        return true 
+    else 
+        is_palindrome_helper(s,index)
+    end
+         
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def is_palindrome_helper(s,index)
+    if s[index]== s[s.length-1-index]
+        is_palindrome_helper(s,index+1)
+    elsif index >= s.length/2
+        return true
+    else 
+        return false
+    end
+end
+
+
+# Time complexity: O(shorter)
+# Space complexity: O(narr+marr)
 def digit_match(n, m)
```

Comment:
You're going to make O(x) recursive calls where x is the number of digits in the smallest number and each time you make strings with `n.digits` and `m.digits`.  So you're almost right, but you're doing it x times.